### PR TITLE
Add unit tests for the assembler and interpreter

### PR DESCRIPTION
**Pull request by tomhea**
The suite previously covered only the stl-library and full assembler+interpreter
integration (the CSV-driven tests in tests/test_fj.py). This adds focused unit
tests for the assembler and interpreter themselves, in a new tests/unit/ package:

- test_expr:        the Expr arithmetic-tree, every math operator, and failures
- test_parser:      lexer number/char/string formats and comment handling
- test_preprocessor:pure macro param-binding, rep-counts, label collectors
- test_assembler:   each language rule compiles + the error/edge cases raise
- test_fjm:         Writer/Reader round-trips (all versions x widths), relative
                    jumps, garbage-handling, validation, and corrupt files
- test_io_devices:  FixedIO bit-ordering/EOF/incomplete-output and BrokenIO
- test_interpreter: each termination cause, the IO/EOF path, last-ops deque
- test_utils:       debug-label round-trip, file helpers, the statistics rule
- test_cli:         the command-line entry-point and version defaulting
- test_quickstart:  end-to-end assemble_and_run across versions and widths

Helpers live in tests/unit/unit_utils.py. The tests are plain functions (no
compile_args/run_args fixtures) so the existing conftest collects them as-is.
The whole suite runs in ~1s and passes black, flake8, mypy --strict and bandit.

### CODE REVIEW

File: tests/unit/test_assembler.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+def test_wflip_zero_value(tmp_path: Path) -> None:
+    compile_and_get_reader(_program('wflip 0x100, 0, l\nl:\n;l\n'), tmp_path)
+
+
+def test_multi_file_macro_sharing(tmp_path: Path) -> None:
+    file_a = MINIMAL_STARTUP + '\ndef shared {\n 0;0\n}\n'
+    file_b = 'startup\nshared\nl:\n;l\n'
+    reader = compile_and_get_reader([file_a, file_b], tmp_path)
+    assert isinstance(reader, Reader)
+
```
Comment (tomhea) [inline tests/unit/test_assembler.py:108]: **Missing language rules: `pad`, `reserve`, `ns` (namespace).** The parser defines statements for all of these (`fj_parser.py:661-693`), but they get no "this compiles" test here. `segment` only appears in the error test below, never as a successful compile. These are first-class language features — please add a one-line smoke test for each (`pad`, `reserve`, `ns`, and a happy-path `segment`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test_assembler.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+# --- error handling / edge cases ---
```
Comment (tomhea) [inline tests/unit/test_assembler.py:110]: **Two high-value error cases are missing here.** The most common real-world assembler errors aren't covered:
- **Macro called with the wrong number of arguments.**
- **Duplicate label definition.**

Both are user-facing failure modes a refactor could break silently. Please add a `pytest.raises` test for each.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test_assembler.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+
+def test_syntax_error_raises(tmp_path: Path) -> None:
+    with pytest.raises(FlipJumpParsingException):
+        assemble_to_path('this is not valid fj $$$ {{{', tmp_path)
+
+
+def test_undefined_label_raises(tmp_path: Path) -> None:
+    with pytest.raises(FlipJumpException):
+        assemble_to_path(_program(';undefined_label\n'), tmp_path)
```
Comment (tomhea) [inline tests/unit/test_assembler.py:120]: Asserting the base `FlipJumpException` is too broad — the code actually raises the specific `FlipJumpAssemblerException` for an undefined label (verified). The base class would also be satisfied by a *parsing* error, so this could pass for the wrong reason. Please tighten to `FlipJumpAssemblerException` to match the style of the other error tests.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test_assembler.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+# --- language rules: each program should assemble into a valid .fjm ---
+
+
+def _program(body: str) -> str:
+    return MINIMAL_STARTUP + '\nstartup\n' + body
+
+
+def test_flip_jump_op(tmp_path: Path) -> None:
+    reader = compile_and_get_reader(_program('a:\n8;a\n'), tmp_path)
+    assert isinstance(reader, Reader)
```
Comment (tomhea) [inline tests/unit/test_assembler.py]: `compile_and_get_reader` is annotated `-> Reader` and always returns one, so `assert isinstance(reader, Reader)` can never fail — it reads like a real check but verifies nothing (same in `test_multi_file_macro_sharing`). Either drop it, or make it meaningful by asserting a specific produced word, e.g. `assert reader.get_word(...) == ...`.

More broadly: these rule tests are pure smoke tests (assembly doesn't throw), so a silent *miscompilation* would pass. A few — at least the arithmetic-bearing `test_wflip_op` and `test_constant_definition_and_use` — would be much stronger by asserting an actual memory word via the returned `Reader`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test_fjm.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+def test_read_write_bit_and_get_word(tmp_path: Path) -> None:
+    fjm_path = _write(tmp_path, 8, FJMVersion.NormalVersion, 0, [0b10110010, 0])
+    reader = Reader(fjm_path)
+
+    assert reader.read_bit(1) is True
+    assert reader.read_bit(0) is False
+    assert reader.get_word(0) == 0b10110010
+
+    reader.write_bit(0, True)
```
Comment (tomhea) [inline tests/unit/test_fjm.py:68]: Two `Reader` branches stay uncovered:
- `write_bit(..., False)` — the bit-*clearing* path (`fjm_reader.py:233-234`). This test only sets a bit to `True`; please also clear one and check the word.
- `get_word` raising when reading past the last bit (`fjm_reader.py:246-247`) — worth a dedicated `pytest.raises(FlipJumpRuntimeMemoryException)`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test_fjm.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    reader = Reader(_write(tmp_path, 16, FJMVersion.NormalVersion, 0, [0, 0]), garbage_handling=GarbageHandling.Stop)
+    with pytest.raises(FlipJumpRuntimeMemoryException) as exc_info:
+        reader.get_word(0x4000)
+    assert exc_info.value.memory_address is not None
+
+
+def test_garbage_handling_continue_returns_zero(tmp_path: Path) -> None:
+    reader = Reader(
+        _write(tmp_path, 16, FJMVersion.NormalVersion, 0, [0, 0]), garbage_handling=GarbageHandling.Continue
+    )
```
Comment (tomhea) [inline tests/unit/test_fjm.py:112]: `assert ... == 0` silently depends on `_new_garbage_val()` returning 0 (`fjm_consts.py:72`). If that constant ever changes, this breaks for a non-obvious reason. A short comment noting the dependency (like the nice one in `test_zeros_boundaries_lazy_init`) would make the intent clear.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by tomhea — COMMENTED** [review]
Reviewed the unit tests against the actual source — they pass (171), run in ~1s, and accurately reflect the code's behavior. I spot-checked the non-obvious assertions (the `register_op` 2*w flip/jump math, `FixedIO`'s LSB-first ordering, the `<HHQQ` header layout in the unsupported-version test, and the v2/v3 relative-jump transparency) and they're all correct. Nice, well-structured addition that closes a real gap.

A few things to fix before merge, left as inline comments. The ones I'd treat as blocking:

1. **Missing language-rule coverage** — `pad`, `reserve`, `ns`, and a happy-path `segment` have no compile test (`test_assembler.py`).
2. **Missing common assembler errors** — wrong macro arg count, and duplicate label (`test_assembler.py`).
3. **`test_undefined_label_raises` asserts the base `FlipJumpException`** instead of the specific `FlipJumpAssemblerException` the code actually raises.

Nice-to-haves (also inline): tautological `isinstance(reader, Reader)` asserts + strengthening a couple of smoke tests to check real memory words; two uncovered `Reader` branches (`write_bit(False)`, `get_word` past last bit); and a comment on the garbage-val `== 0` dependency.

Not blocking / intentionally skipped looks fine: `StandardIO` (interactive) and the `SlowRead`/`OnlyWarning` garbage modes.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by tomhea — COMMENTED** [review]
All review comments addressed correctly — verified each fix against source, not just that the suite is green (178 passed, ~1.4s). (Can't click Approve since it's my own PR, but this is a 👍 from my side.)

**Blocking items — all resolved:**
1. ✅ Added `pad`, `reserve`, happy-path `segment`, and `ns` rule tests. `test_namespace` even asserts the real flip-word (`5`), so it's more than a smoke test.
2. ✅ Added `test_wrong_macro_argument_count_raises` and `test_duplicate_label_raises`. Confirmed both raise the specific `FlipJumpPreprocessorException` for the intended reason, not an incidental error.
3. ✅ `test_undefined_label_raises` tightened to `FlipJumpAssemblerException`.

**Nice-to-haves — also done:**
4. ✅ Replaced the no-op `isinstance` asserts with real memory-word checks (`flip;jump`, constant, `wflip`, multi-file). The `FIRST_OP_FLIP_WORD`/`JUMP_WORD = 4/5` constants (derived from the 2-op `MINIMAL_STARTUP`) are a clean way to express this.
5. ✅ `write_bit(..., False)` clearing path covered, and `test_get_word_past_last_bit_raises` correctly builds an unaligned address on the last word to hit the `fjm_reader.py:246-247` guard.
6. ✅ Garbage-value `== 0` dependency now documented.

Nothing done wrong, no regressions. LGTM 🚀

---
_Generated by [Claude Code](https://claude.ai/code)_
